### PR TITLE
xds: add hashCode and equals back to SslContextProviderSupplier [backport to 1.41.x] (#8528)

### PR DESCRIPTION
**Pull request by sanjaypujare**
backport of #8528

The xdsClient uses equals() for duplicate detection/suppression so a proper equals() (and also hashCode()) is needed. sslContextProvider and shutdown are excluded because these are implementation instances instead of pure data. This will be replaced by a more permanent solution later (such as removing the construction of implementation instances by XdsClient)

### CODE REVIEW

**Review by ejona86 — APPROVED** [review]
We typically delay on the backport until the original PR is approved or merged, since now you'll need to also copy the changes to this backport.

**Comment by sanjaypujare** [comment]
> We typically delay on the backport until the original PR is approved or merged, since now you'll need to also copy the changes to this backport.

Yes, typically I do wait until the master PR is merged and *then* cherry-pick the commit from the upstream/master. However in this case the change was small and I was kind of confident it will be approved/merged as is.

There were no changes in the actual commit - just the description of the PR which I am going to update in the PR. What other changes do I need to backport?

**Comment by ejona86** [comment]
> There were no changes in the actual commit - just the description of the PR which I am going to update in the PR

That description change should have been in the commit as well. See my comment on the original PR.

**Comment by sanjaypujare** [comment]
> > There were no changes in the actual commit - just the description of the PR which I am going to update in the PR
> 
> That description change should have been in the commit as well. See my comment on the original PR.

Oh, okay. I'll update the commit for this one at least.

**Comment by ejona86** [comment]
> Oh, okay. I'll update the commit for this one at least.

Don't bother. The commit on master is much more important for the future than this one and it is very helpful if the commits are identical for a backport.

**Comment by sanjaypujare** [comment]
> > Oh, okay. I'll update the commit for this one at least.
> 
> Don't bother. The commit on master is much more important for the future than this one and it is very helpful if the commits are identical for a backport.

I already updated the commit description. I'll revert it if you prefer that

**Comment by ejona86** [comment]
I think it will be okay to leave the text in the commit.

**Pull request merged**
